zilliqa/src/block_store.rs

```diff
@@ -134,26 +175,36 @@ impl BlockStore {
             {
                 let from = self.requested_view + 1;
                 let to = (self.requested_view + self.batch_size).max(self.highest_known_view);
-                trace!(from, to, "requesting blocks");
-                let message = ExternalMessage::BlockRequest(BlockRequest {
-                    from_view: from,
-                    to_view: to,
-                });
-                let Some(peer) = self.best_peer(from) else {
-                    warn!(from, "no peers to download missing blocks from");
-                    return Ok(());
-                };
-                self.message_sender.send_external_message(peer, message)?;
-
-                let requested_blocks = to - from + 1;
-                self.peers.entry(peer).or_default().requested_blocks += requested_blocks;
-                self.requested_view += requested_blocks;
+                self.request_blocks(from, to)?;
             }
         }
 
         Ok(())
     }
 
+    /// Make a request for a range of blocks. Returns `true` if a request was made and `false` if the request had to be
+    /// buffered because no peers were available.
+    fn request_blocks(&mut self, from: u64, to: u64) -> Result<bool> {
+        let Some(peer) = self.best_peer(from) else {
+            warn!(from, "no peers to download missing blocks from");
+            self.pending_requests.push((from, to));
```

Comment:
this has the effect that pending requests are retried in LIFO order. intentional? (I don't think it greatly matters, but very marginally, presumably more nodes have older blocks and older requests are likely to be for those older blocks?)

Reply:
Correct. I figured it didn't matter since the length of `self.unserviceable_requests` is bounded and once a peer appears, we will send all the requests to it anyway (i.e. the list will be emptied before any of the responses are received).

However, you make a good point about block availability (i.e. its sad if we break early from our `self.unserviceable_requests.pop()` loop because the peer couldn't service a large block number when it could in fact service the lower numbers). I guess I could argue that this case will resolve itself quickly once we learn that said peer can now service a greater block number.

I will add a comment to clarify the current behaviour and suggest changing it in the future if we see problems?